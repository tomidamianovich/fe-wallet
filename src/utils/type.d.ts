export type CurrencyType = {
  ticker: string
  name: string
  symbol: string
  decimals: number
  type: string
  actions: {
    action: string
    enabled: boolean
  }[]
  url_images: {
    image_png: string
    image_svg: string
  }
  color: string | null
  categories: string[]
  is_favorite: boolean
}

export type RateType = {
  ticker: string
  buy_rate: string
  sell_rate: string
  variation: string
}

export type BalanceType = {
  ticker: string
  amount: number
  account_limit: number
}

export type TransactionActionType =
  | 'DEPOSIT'
  | 'WITHDRAWAL'
  | 'SEND'
  | 'RECEIVE'
  | 'BUY'
  | 'SWAP'
export type TransactionStatusType = 'PENDING' | 'FAILURE' | 'SUCCESS'

export type TransactionType = {
  date: Date
  action: TransactionActionType
  ticker: string
  via: string
  status: TransactionStatusType
  comision: number
  amount: number
  total: number
}

export interface transaction {
  ['DEPOSIT']: TransactionActionType
  ['WITHDRAWAL']: TransactionActionType
  ['SEND']: TransactionActionType
  ['RECEIVE']: TransactionActionType
  ['BUY']: TransactionActionType
  ['SWAP']: TransactionActionType
}
export interface transactionTitle {
  DEPOSIT: string
  WITHDRAWAL: string
  SEND: string
  RECEIVE: string
  BUY: string
  SWAP: string
}

export interface transactionStatus {
  ['PENDING']: TransactionStatusType
  ['FAILURE']: TransactionStatusType
  ['SUCCESS']: TransactionStatusType
}

export type Link = {
  name: string
  to: string
  icon: string
}

export type EventSelect = {
  target: {
    value: TransactionActionType
  }
}

export type NavLinkType = link

export type CurrencyState = CurrencyType[]
export type RateState = RateType[]
export type NavLinkState = NavLinkType[]
export type BalanceState = BalanceType[]
export type TransactionState = TransactionType[]

export type CurrencyAction = {
  type: strings
  payload: CurrencyState
}

export type RateAction = {
  type: string
  payload: RateState
}

export type NavLinkAction = {
  type: string
  payload: NavLinkState
}

export type BalanceAction = {
  type: string
  payload: BalanceState
}

export type CurrentBalanceAction = {
  type: string
  payload: BalanceType
}

export type TransactionsAction = {
  type: string
  payload: TransactionState
}

export type BalanceWithCurrencyInfo = BalanceType &
  CurrencyType & { sell_rate: string }

export type CurrencyDispatchType = (args: CurrencyAction) => CurrencyAction
export type RateDispatchType = (args: RateAction) => RateAction
export type NavLinkDispatchType = (args: RateAction) => NavLinkAction
export type BalanceDispatchType = (args: BalanceAction) => BalanceAction
export type TransactionDispatchType = (
  args: TransactionsAction,
) => TransactionsAction

// Combined Types

export type CombinedState = {
  CurrencyReducer: CurrencyState
  RateReducer: RateState
  NavLinkReducer: NavLinkState
  BalanceReducer: BalanceState
  CurrentBalanceReducer: BalanceType
  TransactionsReducer: TransactionState
}

export type CombinedAction = {
  CurrencyReducer: CurrencyAction
  RateReducer: RateAction
  NavLinkReducer: NavLinkAction
  BalanceReducer: BalanceAction
  CurrentBalanceReducer: CurrentBalanceAction
  TransactionsReducer: TransactionsAction
}

export type CombinedDispatchType = (args: CombinedAction) => CombinedAction
