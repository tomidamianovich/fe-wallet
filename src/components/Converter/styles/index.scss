.fe-wallet__content__converter {
  
}