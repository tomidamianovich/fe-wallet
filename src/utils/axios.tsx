import { RateType, CurrencyType } from '../utils/type';
const axios = require('axios').default;

type responseType = {
  data: RateType[] | CurrencyType[]
}
  
const requestHandler = async (url:string): Promise<any> => 
  new Promise((resolve, reject) => 
    axios.get(url)
    .then((response: responseType) => resolve(response?.data))
    .catch((err:any) => reject(err))
  );

export {
  requestHandler,
}