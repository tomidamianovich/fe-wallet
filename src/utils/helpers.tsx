const handleFormatValues = (value:number) => value.toFixed(2);

export {
  handleFormatValues
}