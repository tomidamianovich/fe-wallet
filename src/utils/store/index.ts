import currencies from './currencies';
import cryptoCurrencies from './cryptoCurrencies';
import fiatCurrencies from './fiatCurrencies';
import balances from './balances';
import rates from './rates';

export {
  currencies,
  fiatCurrencies,
  cryptoCurrencies,
  rates,
  balances
}