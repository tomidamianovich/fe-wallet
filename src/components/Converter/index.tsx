import React, { useState } from 'react';
import { CurrencyType, RateType } from '../../utils/type';
import Dropdown from '../Dropdown';
import Icon from '../Icon';
import { withTranslation, WithTranslation } from 'react-i18next';
import './styles/index.scss';

type Props = {
  currencies: CurrencyType[],
  rates: RateType[]
} & WithTranslation;

/*

  Component Name: Converter
  Usages: 
    - Component that allow a user to convert between currencies values.
    - Component that allow a user to know fees in advanced before a purchase or swap.

*/

const Converter: React.FC<Props> = ({
    currencies, 
    rates,
    t
  }) => {
    const [fromCurrencyIndex, setFromCurrencyIndex] = useState<number>(0);
    const [toCurrencyIndex, setToCurrencyIndex] = useState<number>(1);
    const [fromAmount, setFromAmount] = useState<number>(0);
    const [result, setResult] = useState<number>(0);

    const handlerFromCurrency = (e: React.ChangeEvent<HTMLInputElement>) =>  {
      const currencyIndex = parseInt(e.target.value);
      setFromCurrencyIndex(currencyIndex);
    };

    const handlerToCurrency = (e: React.ChangeEvent<HTMLInputElement>) => {
      const currencyIndex = parseInt(e.target.value);
      setToCurrencyIndex(currencyIndex);
    };

    const handlerFromAmount = (e: React.ChangeEvent<HTMLInputElement>) => setFromAmount(parseInt(e.target.value));

    const handlerConversion = () => {
      const currencyTickerFrom = currencies[fromCurrencyIndex]?.ticker;
      const currencyTickerTo = currencies[toCurrencyIndex]?.ticker;
      if (!currencyTickerFrom || !currencyTickerTo) return
      const [{sell_rate}] = rates?.filter(rate => rate.ticker?.includes(currencyTickerFrom) && rate.ticker?.includes('ARS'));
      const [{buy_rate}] = rates?.filter(rate => rate.ticker?.includes(currencyTickerTo) && rate.ticker?.includes('ARS'));
      const money = fromAmount * parseInt(sell_rate);
      const result = buy_rate ? money / parseInt(buy_rate) : 0;
      setResult(result);
    };

    return (
      <div className="fe-wallet__converter">
        <div className="fe-wallet__converter__crypto">
          {currencies && 
            <Dropdown 
              name={t('converter.from.dropdown.name')}
              handler={handlerFromCurrency} 
              options={currencies} 
              selectedIndex={fromCurrencyIndex} 
              ariaLabel={t('converter.from.dropdown.aria_label')}
            />
          }
          <input 
            type="number" 
            className="fe-wallet__converter__input"
            name={t('converter.from.input.name')}
            onChange={handlerFromAmount} 
            defaultValue={0} 
            aria-label={t('converter.from.input.aria_label')}
          />
        </div>
        <div className="fe-wallet__converter__icon">
          <Icon icon="arrow-right" width={24} height={16}/>
        </div>
        <div className="fe-wallet__converter__crypto">
          {currencies && 
            <Dropdown 
              name={t('converter.to.dropdown.name')}
              handler={handlerToCurrency} 
              options={currencies} 
              selectedIndex={toCurrencyIndex} 
              ariaLabel={t('converter.to.dropdown.aria_label')}
            />
          }
          <input 
            type="number"
            className="fe-wallet__converter__input"
            name={t('converter.to.input.name')}
            value={result} 
            readOnly 
            aria-label={t('converter.to.input.aria_label')}
          />
        </div>
        <button 
          onClick={handlerConversion} 
          aria-label={t('converter.button.aria_label')}
          className='fe-wallet__converter__submit'
        >
          {t('converter.button.label')}
        </button>
      </div>  
    )
  }
;

export default withTranslation()(Converter);