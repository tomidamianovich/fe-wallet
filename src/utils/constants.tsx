import { transaction, transactionTitle, transactionStatus } from '../utils/type';

const ENDPOINTS = {
  DEV: {
    BASE_URL: "http://localhost:3001/",
    CURRENCIES: "currencies",
    RATES: "rates",
    BALANCES: "balance",
    EXT: ""
  },
  PROD: {
    BASE_URL: "https://fe-wallet-2022-default-rtdb.firebaseio.com/",
    CURRENCIES: "currencies",
    RATES: "rates",
    BALANCES: "balance",
    EXT: ".json"
  }
}

const PANEL = {
  HEADING: "Mi billetera",
  SUBHEADING: "En tu billetera vas a poder almacenar todas las criptomonedas que compres en ripio.",
}

const TRANSACTIONS:{
  HEADING: string,
  HEADER: string[],
  TYPES: transaction,
  TYPES_TITLES: transactionTitle,
  STATUS: transactionStatus
} = {
  HEADING: "Mis movimientos",
  HEADER: [
    "Fecha",
    "Tipo de Operación",
    "Método",
    "Estado",
    "Comisión",
    "Monto",
  ],
  TYPES: {
    DEPOSIT: "DEPOSIT",
    WITHDRAWAL: "WITHDRAWAL",
    SEND: "SEND",
    RECEIVE: "RECEIVE",
    BUY: "BUY",
    SWAP: "SWAP"
  },
  TYPES_TITLES: {
    DEPOSIT: "Ingreso",
    WITHDRAWAL: "Retiro",
    SEND: "Envio",
    RECEIVE: "Recibo",
    BUY: "Compra",
    SWAP: "Swap"
  },
  STATUS: {
    PENDING: "PENDING",
    FAILURE: "FAILURE",
    SUCCESS: "SUCCESS"
  }
}

const DEFAULT_BALANCE = {
  ticker: "",
  amount: 0,
  account_limit: 0,
  name: "",
  symbol: "",
  decimals: 2,
  type: "",
  actions: [{
    action: "",
    enabled: true
  }],
  url_images: {
    image_png: "",
    image_svg: "",
  },
  color: "",
  categories: [],
  is_favorite: true,
  sell_rate: ""
}

const COMISIONS = {
  BUY: 0.003,
  DEPOSIT: 0,
  WITHDRAWAL: 0.0004,
  SEND: 0,
  RECEIVE: 0,
  SWAP: 0.00004
}

type Link = {
  name: string;
  to: string;
  icon: string;
};

const NAV_LINKS:Link[] = [
  {
    name: 'Panel',
    to: 'panel',
    icon: 'panel'
  },
  {
    name: 'Billetera',
    to: 'wallet',
    icon: 'wallet'
  },
  {
    name: 'Cotizaciones',
    to: 'market-rates',
    icon: 'market-rates'
  },
  {
    name: 'Créditos',
    to: 'credit',
    icon: 'credit'
  },
  {
    name: 'Exchange',
    to: 'exchange',
    icon: 'exchange'
  },
  {
    name: 'Launchpad',
    to: 'launchpad',
    icon: 'launchpad'
  }
];

const CURRENCIES = {
  TYPES: {
    FIAT: "FIAT",
    CRYPTO: "CRYPTO"
  }
};

export {
  ENDPOINTS,
  PANEL,
  TRANSACTIONS,
  DEFAULT_BALANCE,
  COMISIONS,
  NAV_LINKS,
  CURRENCIES
}