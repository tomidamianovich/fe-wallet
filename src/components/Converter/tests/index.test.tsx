import { ComponentType } from 'react'
import { render, screen, fireEvent, waitFor } from '@testing-library/react'
import Converter from '../index'
import { cryptoCurrencies, rates } from '../../../utils/tests'

jest.mock('react-i18next', () => ({
  withTranslation: () => (Component: ComponentType) => {
    Component.defaultProps = {
      ...Component.defaultProps,
      t: (translation: string) => translation,
    }
    return Component
  },
}))

test('renders Converter  with currencies and rates', () => {
  const { asFragment } = render(
    <Converter currencies={cryptoCurrencies} rates={rates} />,
  )
  expect(asFragment()).toMatchSnapshot()
})

test('renders Converter  without currencies and rates', () => {
  const { asFragment } = render(<Converter currencies={[]} rates={[]} />)
  expect(asFragment()).toMatchSnapshot()
})

test('Testing Converter with currencies and rates', async () => {
  render(<Converter currencies={cryptoCurrencies} rates={rates} />)

  const inputFrom = screen.getByRole('spinbutton', {
    name: 'converter.from.input.aria_label',
  })
  const inputTo = screen.getByRole('spinbutton', {
    name: 'converter.to.input.aria_label',
  })
  expect(inputFrom).toBeInTheDocument()
  expect(inputFrom).toHaveValue(0)
  expect(inputTo).toBeInTheDocument()
  expect(inputTo).toHaveValue(0)

  const selectedCryptoFrom = cryptoCurrencies[0].ticker
  const selectedCryptoTo = cryptoCurrencies[1].ticker
  const selectFrom = screen.getByRole('combobox', {
    name: 'converter.from.dropdown.aria_label',
  })
  const selectTo = screen.getByRole('combobox', {
    name: 'converter.to.dropdown.aria_label',
  })
  expect(selectFrom).toBeInTheDocument()
  expect(selectFrom).toHaveValue('0')
  expect(selectTo).toBeInTheDocument()
  expect(selectTo).toHaveValue('1')
  expect(selectFrom).toHaveDisplayValue('BTC')
  expect(selectTo).toHaveDisplayValue('ETH')

  fireEvent.change(inputFrom, { target: { value: '10' } })
  expect(inputFrom).toHaveValue(10)
  await waitFor(() => expect(inputTo).toHaveValue(0))

  const convertirButton = screen.getByRole('button', {
    name: 'converter.button.aria_label',
  })
  expect(convertirButton).toBeInTheDocument()
  fireEvent.click(convertirButton)
  const [{ sell_rate }] = rates.filter(
    rate =>
      rate.ticker?.includes(selectedCryptoFrom) && rate.ticker?.includes('ARS'),
  )
  const [{ buy_rate }] = rates.filter(
    rate =>
      rate.ticker?.includes(selectedCryptoTo) && rate.ticker?.includes('ARS'),
  )
  const money = 10 * parseInt(sell_rate)
  const result = buy_rate ? money / parseInt(buy_rate) : 0
  await waitFor(() => expect(inputTo).toHaveValue(result))
})
