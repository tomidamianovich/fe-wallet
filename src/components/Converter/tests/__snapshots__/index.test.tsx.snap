// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Converter  with currencies and rates 1`] = `
<DocumentFragment>
  <div
    class="fe-wallet__converter"
  >
    <div
      class="fe-wallet__converter__crypto"
    >
      <select
        aria-label="converter.from.dropdown.aria_label"
        class="fe-wallet__dropdown "
        id="converter.from.dropdown.name"
        name="converter.from.dropdown.name"
      >
        <option
          value="0"
        >
          BTC
        </option>
        <option
          value="1"
        >
          ETH
        </option>
        <option
          value="2"
        >
          USDC
        </option>
        <option
          value="3"
        >
          LTC
        </option>
        <option
          value="4"
        >
          DAI
        </option>
        <option
          value="5"
        >
          AXS
        </option>
      </select>
      <input
        aria-label="converter.from.input.aria_label"
        class="fe-wallet__converter__input"
        name="converter.from.input.name"
        type="number"
        value="0"
      />
    </div>
    <div
      class="fe-wallet__converter__icon"
    >
      <svg
        class="icon"
        height="16"
        viewBox="0 0 24 16"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M14.501 2.52607C14.2545 2.26833 14.2546 1.86225 14.501 1.60453L15.8387 0.205879C16.1013 -0.068637 16.5398 -0.0686247 16.8023 0.205906L23.8151 7.53914C24.0616 7.79687 24.0616 8.20292 23.8152 8.46066L16.8023 15.7941C16.5398 16.0686 16.1012 16.0686 15.8387 15.7941L14.501 14.3952C14.2546 14.1375 14.2546 13.7315 14.501 13.4737L18.1522 9.6558L0.666666 9.6558C0.298476 9.6558 0 9.35732 0 8.98913V7.01087C0 6.64268 0.298476 6.3442 0.666666 6.3442L18.1522 6.3442L14.501 2.52607Z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <div
      class="fe-wallet__converter__crypto"
    >
      <select
        aria-label="converter.to.dropdown.aria_label"
        class="fe-wallet__dropdown "
        id="converter.to.dropdown.name"
        name="converter.to.dropdown.name"
      >
        <option
          value="0"
        >
          BTC
        </option>
        <option
          value="1"
        >
          ETH
        </option>
        <option
          value="2"
        >
          USDC
        </option>
        <option
          value="3"
        >
          LTC
        </option>
        <option
          value="4"
        >
          DAI
        </option>
        <option
          value="5"
        >
          AXS
        </option>
      </select>
      <input
        aria-label="converter.to.input.aria_label"
        class="fe-wallet__converter__input"
        name="converter.to.input.name"
        readonly=""
        type="number"
        value="0"
      />
    </div>
    <button
      aria-label="converter.button.aria_label"
      class="fe-wallet__converter__submit"
    >
      converter.button.label
    </button>
  </div>
</DocumentFragment>
`;

exports[`renders Converter  without currencies and rates 1`] = `
<DocumentFragment>
  <div
    class="fe-wallet__converter"
  >
    <div
      class="fe-wallet__converter__crypto"
    >
      <select
        aria-label="converter.from.dropdown.aria_label"
        class="fe-wallet__dropdown "
        id="converter.from.dropdown.name"
        name="converter.from.dropdown.name"
      />
      <input
        aria-label="converter.from.input.aria_label"
        class="fe-wallet__converter__input"
        name="converter.from.input.name"
        type="number"
        value="0"
      />
    </div>
    <div
      class="fe-wallet__converter__icon"
    >
      <svg
        class="icon"
        height="16"
        viewBox="0 0 24 16"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M14.501 2.52607C14.2545 2.26833 14.2546 1.86225 14.501 1.60453L15.8387 0.205879C16.1013 -0.068637 16.5398 -0.0686247 16.8023 0.205906L23.8151 7.53914C24.0616 7.79687 24.0616 8.20292 23.8152 8.46066L16.8023 15.7941C16.5398 16.0686 16.1012 16.0686 15.8387 15.7941L14.501 14.3952C14.2546 14.1375 14.2546 13.7315 14.501 13.4737L18.1522 9.6558L0.666666 9.6558C0.298476 9.6558 0 9.35732 0 8.98913V7.01087C0 6.64268 0.298476 6.3442 0.666666 6.3442L18.1522 6.3442L14.501 2.52607Z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <div
      class="fe-wallet__converter__crypto"
    >
      <select
        aria-label="converter.to.dropdown.aria_label"
        class="fe-wallet__dropdown "
        id="converter.to.dropdown.name"
        name="converter.to.dropdown.name"
      />
      <input
        aria-label="converter.to.input.aria_label"
        class="fe-wallet__converter__input"
        name="converter.to.input.name"
        readonly=""
        type="number"
        value="0"
      />
    </div>
    <button
      aria-label="converter.button.aria_label"
      class="fe-wallet__converter__submit"
    >
      converter.button.label
    </button>
  </div>
</DocumentFragment>
`;
